### DISCUSSION TITLE
Clarification Needed on num_depot Usage in MDCPDP Environment

### DISCUSSION THREAD

**Discussion by Moonbohoon** [ai4co/rl4co#220] [General]
I'm currently working with the `MDCPDPEnv` environment from this repository and encountered an issue regarding the usage of `num_depot`. In the `_step` and `_get_reward` function of `env.py` and `render.py`, num_depot is derived from `td["capacity"].shape[-1]`, which seems unusual to me. Here's the relevant code snippet:

``` python
num_depot = td["capacity"].shape[-1]
```

I believe it should be derived from `depot`, not `capacity`. Is this understanding correct? 
Any guidance or confirmation on this issue would be greatly appreciated. Thank you!
Best regards,

**Comment by Moonbohoon** [comment]
any update in here?

**Comment by 699Felix** [comment]
I modified the code in class MDCPDPGenerator:
`
capacity = torch.randint(
            self.min_capacity,
            self.max_capacity + 1,
            # size=(*batch_size, 1),
            size=(*batch_size, self.num_depot),
        )
`
then you can see more depots not only one
but since here i cannot start the trainning of the MVMoE_POMO model……

**Reply by fedebotu** [reply]
Hi Felix, sorry for the veeery late response. Will keep an eye out in discussions from now on. 

Did you manage to fix the above?  Probably, the problem above was due to the way we handle the first starting node in POMO, which should be modified for this environment :eyes:

**Reply by 699Felix** [reply]
> Hi Felix, sorry for the veeery late response. Will keep an eye out in discussions from now on. 
> 
> Did you manage to fix the above?  Probably, the problem above was due to the way we handle the first starting node in POMO, which should be modified for this environment :eyes: 

Hi, I've not been working on this problem for a long time,later I will take a try,thanks for your effort!

**Reply by fedebotu** [reply]
Sure, feel free to let us know if any help is needed!

**Comment by fedebotu** [comment]
Hi @Moonbohoon and @699Felix !

(I'm sorry that we have completely missed this discussion for a long time now :sweat_smile: )

We have made some updates to the environment and model, that we report here:


These fixes are based on feedback from @Moonbohoon in #220 (which was due quite some time ago :sweat_smile: ) and updates to the parallel autoregressive counterpart of this environment in [PARCO](https://github.com/ai4co/parco).

Now `num_depot` is replaced with `num_agents`; each agent has its own starting node, which can be either in the same location (as in `single` depot) or with different locations (`multiple` depot). 


Below is an example to instantiate a model with the new env:

```python
import torch
from rl4co.utils.trainer import RL4COTrainer

from rl4co.envs.routing.mdcpdp.env import MDCPDPEnv, MDCPDPGenerator
from rl4co.models.zoo.am.policy import AttentionModelPolicy
from rl4co.models.nn.env_embeddings.init import MDCPDPInitEmbedding
from rl4co.models.nn.env_embeddings.context import MDCPDPContext

# Greedy rollouts over trained model
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

embed_dim = 128
policy = AttentionModelPolicy(
            env_name="mdcpdp",
            init_embedding=MDCPDPInitEmbedding(embed_dim),
            context_embedding=MDCPDPContext(embed_dim),
            embed_dim=embed_dim,
        ).to(device)


generator = MDCPDPGenerator(min_capacity=2, max_capacity=3, num_agents=10, num_loc=60, depot_mode="multiple") # or set to "single" for M agents with the same starting location
env_ar = MDCPDPEnv(generator, problem_mode="open")

td_gen_ar = env_ar.generator(26)
td_reset_ar = env_ar.reset(td_gen_ar.clone()).to(device)

# Inference
with torch.inference_mode():
    out_ar = policy(td_reset_ar.clone(), env_ar, decode_type="greedy")

# Plotting (untrained!)
actions_ar = out_ar["actions"]# .reshaape(td_init.shape[0], -1)
print("Average tour length: {:.2f}".format(-out_ar['reward'].mean().item()))
for i in range(2):
    print(f"Tour {i} length: {-out_ar['reward'][i].item():.2f}")
    env_ar.render(td_reset_ar[i], actions_ar[i].cpu())
```
